packages/react-router-dom/__tests__/special-characters-test.tsx

```diff
@@ -0,0 +1,712 @@
+            search: "",
+            hash: "",
+          },
+          { "*": `foo/bar${char}` }
+        );
+      }
+    });
+
+    it("handles special chars in search params", async () => {
+      for (let charDef of specialChars) {
+        let { char, searchChar } = charDef;
+        await testParamValues(`/path?key=${char}`, "Static Route", {
+          pathname: `/path`,
+          search: `?key=${searchChar}`,
+          hash: "",
+        });
+      }
+    });
+
+    it("handles special chars in hash values", async () => {
+      for (let charDef of specialChars) {
+        let { char, hashChar } = charDef;
+        await testParamValues(`/path#hash-${char}`, "Static Route", {
+          pathname: `/path`,
+          search: "",
+          hash: `#hash-${hashChar}`,
+        });
+      }
+    });
+  });
+
+  describe("when matching as part of the defined route path", () => {
+    async function assertRouteMatch(
+      char,
+      path: string,
+      expectedHeading: string,
+      expectedLocation: Omit<Location, "state" | "key">,
+      expectedParams = {}
+    ) {
+      // Need to use our own custom DOM in order to get a working history
+      const dom = new JSDOM(`<!DOCTYPE html><html><body></body></html>`, {
+        url: "https://remix.run/",
+      });
+      let testWindow = dom.window as unknown as Window;
+      testWindow.history.replaceState(null, "", path);
+
+      let renderedUseLocation: Omit<Location, "state" | "key"> | null = null;
+      let renderedParams: Params<string> | null = null;
+
+      function CaptureLocation() {
+        let location = {
+          ...useLocation(),
+          state: undefined,
+          key: undefined,
+        };
+        let params = useParams();
+        renderedUseLocation = location;
+        renderedParams = params;
+        return (
+          <>
+            <p>{location.pathname}</p>
+            <Link to="/reset">Link to reset</Link>
+          </>
+        );
+      }
+
+      function Root() {
+        return (
+          <>
+            <h1>Matched Root</h1>
+            <CaptureLocation />
+          </>
+        );
+      }
+
+      function StaticNested() {
+        return (
+          <>
+            <h1>Matched Static Nested</h1>
+            <CaptureLocation />
+          </>
+        );
+      }
+
+      function ParamNested() {
+        return (
+          <>
+            <h1>Matched Param Nested</h1>
+            <CaptureLocation />
+          </>
+        );
+      }
+
+      let routeElements = (
+        <>
+          <Route path={`/${char}`} element={<Root />} />
+          <Route path="/nested">
+            <Route path={char} element={<StaticNested />} />
+          </Route>
+          <Route path="/:param">
+            <Route path={char} element={<ParamNested />} />
+          </Route>
+          <Route path="/reset" element={<Link to={path}>Link to path</Link>} />
+          <Route path="*" element={<h1>Not Found</h1>} />
+        </>
+      );
```

Comment:
Test that we can define and match against route paths containing special characters